File: S4-Systems-for-Design-and-Impl/Labs/Lab5/ProiectJavaNew/src/main/java/service/Service.java
package service;

import domain.Reservation;
import domain.ReservationManager;
import domain.Seat;
import domain.Trip;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.interfaces.IReservationManagerRepository;
import repository.interfaces.IReservationRepository;
import repository.interfaces.ISeatRepository;
import repository.interfaces.ITripRepository;

import java.util.ArrayList;
import java.util.List;

public class Service {

    private static final Logger logger = LogManager.getLogger();

    private IReservationManagerRepository reservationManagerRepository;
    private IReservationRepository reservationRepository;
    private ITripRepository tripRepository;
    private ISeatRepository seatRepository;

    public Service(IReservationManagerRepository reservationManagerRepository, IReservationRepository reservationRepository, ITripRepository tripRepository, ISeatRepository seatRepository) {
        this.reservationManagerRepository = reservationManagerRepository;
        this.reservationRepository = reservationRepository;
        this.tripRepository = tripRepository;
        this.seatRepository = seatRepository;
    }

    public ReservationManager addReservationManager(ReservationManager reservationManager) {
        return reservationManagerRepository.save(reservationManager).orElseThrow(() -> new RuntimeException("Error adding reservation manager"));
    }

    public ReservationManager deleteReservationManager(Long id) {
        return reservationManagerRepository.delete(id).orElseThrow(() -> new RuntimeException("Error deleting reservation manager"));
    }

    public ReservationManager updateReservationManager(ReservationManager reservationManager) {
        return reservationManagerRepository.update(reservationManager).orElseThrow(() -> new RuntimeException("Error updating reservation manager"));
    }

    public Iterable<ReservationManager> getAllReservationManagers() {
        return reservationManagerRepository.findAll();
    }

    public ReservationManager searchReservationManagerById(Long id) {
        return reservationManagerRepository.findOne(id).orElseThrow(() -> new RuntimeException("Error getting reservation manager"));
    }

    public ReservationManager searchReservationManagerByName(String name) {
        //logger.traceEntry();
        Iterable<ReservationManager> reservationManagers = reservationManagerRepository.findAll();
        for (ReservationManager reservationManager : reservationManagers) {
            if (reservationManager.getName().equals(name)) {
                return reservationManager;
            }
        }
        logger.traceExit();
        return null;
    }

    public Trip addTrip(Trip trip) {
        return tripRepository.save(trip).orElseThrow(() -> new RuntimeException("Error adding trip"));
    }

    public Trip deleteTrip(Long id) {
        return tripRepository.delete(id).orElseThrow(() -> new RuntimeException("Error deleting trip"));
    }

    public Trip updateTrip(Trip trip) {
        return tripRepository.update(trip).orElseThrow(() -> new RuntimeException("Error updating trip"));
    }

    public Iterable<Trip> getAllTrips() {
        return tripRepository.findAll();
    }

    public Trip searchTripById(Long id) {
        return tripRepository.findOne(id).orElseThrow(() -> new RuntimeException("Error getting trip"));
    }

    public Reservation addReservation(Long tripId, List<Long> seats, String clientName) {
        logger.traceEntry();
        Trip trip = searchTripById(tripId);
        trip.setNoOfAvailableSeats(trip.getNoOfAvailableSeats() - seats.size());
        trip = tripRepository.update(trip).orElseThrow(() -> new RuntimeException("Error updating trip"));
        Reservation reservation = new Reservation(trip, seats, clientName);
        reservation = reservationRepository.save(reservation).orElseThrow(() -> new RuntimeException("Error adding reservation"));
        logger.traceExit();
        return reservation;
    }

    public List<Reservation> getReservationsByTrip(Long tripId) {
        //logger.traceEntry();
        Iterable<Reservation> reservations = reservationRepository.findAll();
        List<Reservation> reservationsList = new ArrayList<>();
        for (Reservation reservation : reservations) {
            if (reservation.getTrip().getId().equals(tripId)) {
                reservationsList.add(reservation);
            }
        }
        logger.traceExit();
        return reservationsList;
    }

    public List<Seat> getSeatsByReservation(Long reservationId) {
        //logger.traceEntry();
        Iterable<Seat> seats = seatRepository.findAll();
        List<Seat> seatsList = new ArrayList<>();
        for (Seat seat : seats) {
            if (seat.getReservation().getId().equals(reservationId)) {
                seatsList.add(seat);
            }
        }
        //logger.traceExit();
        return seatsList;
    }
}

File: S4-Systems-for-Design-and-Impl/Labs/Lab6/ProiectCSharpNew/ProiectRezervariCurse/service/Service.cs
using System;
using System.Collections.Generic;
using log4net;
using ProiectRezervariCurse.domain;
using ProiectRezervariCurse.repository.interfaces;

namespace ProiectRezervariCurse.service
{
    public class Service
    {
        private static readonly ILog log = LogManager.GetLogger(typeof(Service));

        private IReservationManagerRepository reservationManagerRepository;
        private IReservationRepository reservationRepository;
        private ITripRepository tripRepository;

        public Service(IReservationManagerRepository reservationManagerRepository, IReservationRepository reservationRepository, ITripRepository tripRepository)
        {
            this.reservationManagerRepository = reservationManagerRepository;
            this.reservationRepository = reservationRepository;
            this.tripRepository = tripRepository;
        }
        
        public void AddReservationManager(ReservationManager reservationManager)
        {
            log.Info("Adding reservation manager " + reservationManager);
            reservationManagerRepository.Save(reservationManager);
        }

        public ReservationManager FindReservationManagerByName(string name)
        {
            log.Info("Finding reservation manager by name " + name);
            IEnumerable<ReservationManager> reservationManagers = reservationManagerRepository.FindAll();
            foreach (var reservationManager in reservationManagers)
            {
                if (reservationManager.Name.Equals(name))
                {
                    return reservationManager;
                }
            }
            return null;
        }

        public IEnumerable<Trip> getAllTrips()
        {
            log.Info("Getting all trips");
            return tripRepository.FindAll();
        }
        
        public Trip FindTripById(long id)
        {
            log.Info("Finding trip by id " + id);
            Trip trip = tripRepository.FindOne(id);
            if (trip != null)
            {
                return trip;
            }
            throw new Exception("Trip not found");
        }
        
        public Reservation AddReservation(long tripId, List<long> noOfSeats, string clientName)
        {
            log.Info("Adding reservation ");
            Trip trip = FindTripById(tripId);
            trip.NoOfAvailableSeats -= noOfSeats.Count;
            trip = tripRepository.Update(trip) ?? throw new Exception("Error updating trip");
            Reservation reservation = new Reservation(trip, noOfSeats, clientName);
            reservation = reservationRepository.Save(reservation) ?? throw new Exception("Error saving reservation");
            log.Info("Exiting AddReservation");
            return reservation;
        }
        
        public List<Reservation> getReservationsByTripId(long tripId)
        {
            log.Info("Getting reservations by trip id " + tripId);
            List<Reservation> reservations = new List<Reservation>();
            IEnumerable<Reservation> allReservations = reservationRepository.FindAll();
            foreach (var reservation in allReservations)
            {
                if (reservation.Trip.Id == tripId)
                {
                    reservations.Add(reservation);
                }
            }
            log.Info("Exiting getReservationsByTripId");
            return reservations;
        }
    }
}